contributed_definitions/nyaml/NXactuator.yaml

```diff
@@ -0,0 +1,198 @@
+category: base
+doc: |
+  An actuator used to control an external condition.
+  
+  The condition itself is described in :ref:`NXenvironment`.
+type: group
+NXactuator(NXobject):
+  model:
+    doc: |
+      Actuator identification code/model number
+  name:
+    doc: |
+      Name of the actuator
+  short_name:
+    doc: |
+      Short name of actuator used e.g. on monitor display program
+  attached_to:
+    doc: |
+      where actuator is attached to
+  physical_quantity:
+    doc: |
+      Name for the physical quantity effected by actuation
+      
+      Examples:
+      temperature | pH | magnetic_field | electric_field | current | conductivity | resistance | voltage |
+      pressure | flow | stress | strain | shear | surface_pressure
+  type:
+    doc: |
+      The type of hardware used for the actuation.
+      
+      Examples (suggestions but not restrictions):
+
+      :Temperature: laser | gas lamp | filament | resistive
+      :Pressure: anvil cell
+      :Voltage: potentiostat
+  OUTPUT(NX_FLOAT):
+    unit: NX_ANY
+    doc: |
+      Any output that the actuator produces.
+      For example, a heater can have the field heater_power(NX_FLOAT).
+  OUTPUT_log(NXlog):
+    doc: |
+      Time history of actuator outputs.
+  (NXpid):
+    doc: |
+      If the actuator is PID-controlled, the settings of the PID controller can be
+      stored here.
+    setpoint(NX_FLOAT):
+      unit: NX_ANY
+      doc: |
+        Nominal actuator setpoint.
+        Can be a scalar or a vector (of [n] actuations).
```

Comment:
> I am puzzled about the assumption that the concept can be a scalar or vector, I thought one has to state the dimensionality in the base class. If no dim is given it is always a scalar until I specialize the concept in an appdef to something where the dim is a vector. Never used this:
> Alas, I agree that when we use inheritance really rigorously, it is beneficial to have the dimensions not be restricted or "being required to state them always at all necessarily already at the base class level"
> Point is this docstring leaves me wondering if we have defined a crystal-clear best practice where the dimensions should be defined.

If the dimensionality is not specified it can have any, including being a scalar, vector or higher dimensional tensor. We intendedly use this in NXmpes to guarantee flexibility.

Should also not restricted here because we specifically want to support both, a scalar _and_ a vector.

> On another point - as seen here in this concept - one could argue that it is sufficient in the base class to say I have a concept labelled by symbol NXactuator/NXpid/setpoint which is a "Nominal ...." without even constraining the instance data of the concept to be NX_FLOAT, in which case namely it must not be an NX_INT unless I specialize..
> NX_ANY essentially states just the default value and strictly speaking even that one does only restrict in so far that the instance value is for sure neither dimensionless nor unitless i.e. a shallow restriction.

I'm not sure if I understand what you are trying to say. For me this contains two points: the field not necessarily being `NX_FLOAT` and specialising it when using it in an appedef. For this case I don't know what else the type of the field should be, because we cannot specify a field _without_ any type in nexus.
The second point is related to this having the unit `NX_ANY`. I kind of agree to your point that it is not particular necessary to have this, but the thing we are doing by specifying this here is that the value should have a value attached, which I think is important for a setpoint value.

Reply:
The point I have with this is: I was (incorrectly) assuming instances of field(NXDL_TYPE_CATEGORY) != NX_CHAR need to have a dimensions. But after checking the NXDL spec indeed this is not required per se.
So yes you can define a field(SOME_CATEGORY) without dim like here and it has then exactly the effect that you can pass scalar, vectors. If you state a dimensions though that becomes part of the concept (at least in our inheritance-capable design) and in this case field(SOME_CATEGORY) you have to make a choice.

After seeing it how it is used here in NXmpes I think it might overall be a good idea but keep in mind that things eventually for the purpose insufficiently strong constraint may fire back when you have a consuming application. Typical case is computational geometry just stating that it is an NX_FLOAT is not sufficient using e.g. an float32 where float64 is needed to signify that two primitives overlap given that discretization can have substantial effect on the correct functioning of the code
but for here, yes, approved.